File: src/main/java/com/healthpal/api/repository/UserRepository.java
package com.healthpal.api.repository;

import com.healthpal.api.model.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

/**
 * UserRepository - Data Access Layer for User entity
 * 
 * This repository is part of the MVC Layered Architecture (Repository/DAO Layer)
 * It extends JpaRepository to provide CRUD operations and custom queries for User entity
 * 
 * @author HealthPal Team - Person A (Authentication + Admin + Core Architecture)
 * @version 1.0
 */
@Repository
public interface UserRepository extends JpaRepository<User, Long> {

    /**
     * Find a user by username
     * @param username The username to search for
     * @return Optional containing the user if found
     */
    Optional<User> findByUsername(String username);

    /**
     * Find a user by email
     * @param email The email to search for
     * @return Optional containing the user if found
     */
    Optional<User> findByEmail(String email);

    /**
     * Find a user by username or email
     * @param username The username to search for
     * @param email The email to search for
     * @return Optional containing the user if found
     */
    Optional<User> findByUsernameOrEmail(String username, String email);

    /**
     * Check if a username exists
     * @param username The username to check
     * @return true if the username exists, false otherwise
     */
    Boolean existsByUsername(String username);

    /**
     * Check if an email exists
     * @param email The email to check
     * @return true if the email exists, false otherwise
     */
    Boolean existsByEmail(String email);

    /**
     * Find all active users
     * @return List of active users
     */
    @Query("SELECT u FROM User u WHERE u.isActive = true ORDER BY u.createdAt DESC")
    List<User> findAllActiveUsers();

    /**
     * Find all admin users
     * @return List of admin users
     */
    @Query("SELECT u FROM User u WHERE u.isAdmin = true ORDER BY u.createdAt DESC")
    List<User> findAllAdmins();

    /**
     * Find users by email verification status
     * @param isVerified The verification status
     * @return List of users with the specified verification status
     */
    @Query("SELECT u FROM User u WHERE u.isEmailVerified = :isVerified ORDER BY u.createdAt DESC")
    List<User> findByEmailVerificationStatus(@Param("isVerified") Boolean isVerified);

    /**
     * Find users by a specific role
     * @param roleName The name of the role
     * @return List of users with the specified role
     */
    @Query("SELECT u FROM User u JOIN u.roles r WHERE r.name = :roleName ORDER BY u.createdAt DESC")
    List<User> findUsersByRole(@Param("roleName") String roleName);
}

File: repository/userRepository.js
const User = require('../models/User');
const logger = require('../utils/logger');

class UserRepository {
    async save(userData) {
        try {
            const user = new User(userData);
            return await user.save();
        } catch (error) {
            logger.error(`Error saving user: ${error.message}`);
            throw error;
        }
    }

    async findById(id) {
        try {
            return await User.findById(id)
                .populate('roles', 'name description')
                .select('-password');
        } catch (error) {
            logger.error(`Error finding user by ID ${id}: ${error.message}`);
            throw error;
        }
    }

    async findAll() {
        try {
            return await User.find()
                .populate('roles', 'name description')
                .select('-password')
                .sort({ createdAt: -1 });
        } catch (error) {
            logger.error(`Error finding all users: ${error.message}`);
            throw error;
        }
    }

    async findByUsername(username) {
        try {
            return await User.findOne({ username })
                .populate('roles', 'name description')
                .select('+password');
        } catch (error) {
            logger.error(`Error finding user by username ${username}: ${error.message}`);
            throw error;
        }
    }

    async findByEmail(email) {
        try {
            return await User.findOne({ email })
                .populate('roles', 'name description')
                .select('+password');
        } catch (error) {
            logger.error(`Error finding user by email ${email}: ${error.message}`);
            throw error;
        }
    }

    async findByUsernameOrEmail(username, email) {
        try {
            return await User.findOne({
                $or: [
                    { username },
                    { email }
                ]
            })
            .populate('roles', 'name description')
            .select('+password');
        } catch (error) {
            logger.error(`Error finding user by username/email ${username}/${email}: ${error.message}`);
            throw error;
        }
    }

    async existsByUsername(username) {
        try {
            const count = await User.countDocuments({ username });
            return count > 0;
        } catch (error) {
            logger.error(`Error checking username existence ${username}: ${error.message}`);
            throw error;
        }
    }

    async existsByEmail(email) {
        try {
            const count = await User.countDocuments({ email });
            return count > 0;
        } catch (error) {
            logger.error(`Error checking email existence ${email}: ${error.message}`);
            throw error;
        }
    }

    async findAllActiveUsers() {
        try {
            return await User.find({ isActive: true })
                .populate('roles', 'name description')
                .select('-password')
                .sort({ createdAt: -1 });
        } catch (error) {
            logger.error(`Error finding all active users: ${error.message}`);
            throw error;
        }
    }

    async findAllAdmins() {
        try {
            return await User.find({ isAdmin: true })
                .populate('roles', 'name description')
                .select('-password')
                .sort({ createdAt: -1 });
        } catch (error) {
            logger.error(`Error finding all admins: ${error.message}`);
            throw error;
        }
    }

    async findByEmailVerificationStatus(isVerified) {
        try {
            return await User.find({ isEmailVerified: isVerified })
                .populate('roles', 'name description')
                .select('-password')
                .sort({ createdAt: -1 });
        } catch (error) {
            logger.error(`Error finding users by email verification status ${isVerified}: ${error.message}`);
            throw error;
        }
    }

    async findUsersByRole(roleName) {
        try {
            return await User.find()
                .populate({
                    path: 'roles',
                    match: { name: roleName },
                    select: 'name description'
                })
                .select('-password')
                .sort({ createdAt: -1 })
                .then(users => users.filter(user => user.roles && user.roles.length > 0));
        } catch (error) {
            logger.error(`Error finding users by role ${roleName}: ${error.message}`);
            throw error;
        }
    }

    async update(id, updateData) {
        try {
            return await User.findByIdAndUpdate(
                id,
                { ...updateData, updatedAt: Date.now() },
                { new: true, runValidators: true }
            )
            .populate('roles', 'name description')
            .select('-password');
        } catch (error) {
            logger.error(`Error updating user ${id}: ${error.message}`);
            throw error;
        }
    }

    async delete(id) {
        try {
            const result = await User.findByIdAndDelete(id);
            return result !== null;
        } catch (error) {
            logger.error(`Error deleting user ${id}: ${error.message}`);
            throw error;
        }
    }

    async deleteAll() {
        try {
            const result = await User.deleteMany({});
            return result.deletedCount;
        } catch (error) {
            logger.error(`Error deleting all users: ${error.message}`);
            throw error;
        }
    }

    async count(filter = {}) {
        try {
            return await User.countDocuments(filter);
        } catch (error) {
            logger.error(`Error counting users: ${error.message}`);
            throw error;
        }
    }

    async findWithFields(filter, fields) {
        try {
            return await User.find(filter)
                .select(fields)
                .populate('roles', 'name description')
                .sort({ createdAt: -1 });
        } catch (error) {
            logger.error(`Error finding users with fields: ${error.message}`);
            throw error;
        }
    }
}

module.exports = new UserRepository();